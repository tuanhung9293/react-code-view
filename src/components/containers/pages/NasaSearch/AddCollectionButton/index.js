import React from 'react'
import PrimaryButton from 'components/ui/buttons/PrimaryButton'

const AddCollectionButton = ({itemData}) => {
  return (
    <PrimaryButton outlined wide to={`/nasa-search/add-media/${itemData.nasaId}`}>
      Add to NASA Collection
    </PrimaryButton>
  )
}

export default AddCollectionButton
