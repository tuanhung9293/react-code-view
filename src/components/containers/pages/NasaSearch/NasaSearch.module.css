.link {
    padding: 15px 0;
}

.loadMore {
    padding-top: 25px;
    padding-bottom: 40px;
    text-align: center;
}

.loading {
    position: relative;
    padding-top: 25px;
    padding-bottom: 40px;
}