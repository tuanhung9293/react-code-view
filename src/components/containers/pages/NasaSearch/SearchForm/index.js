import React from 'react'
import {connect} from 'react-redux'
import {withRouter} from 'react-router-dom'
import {objectToUrlParams, urlParamsToObject} from 'services//transformers'
import {searchNasaMedia} from 'state/actions/mediaAction'
import Input from 'components/ui/forms/Input'
import MediaTypeCheckbox from 'components/ui/forms/MediaTypeCheckbox'
import styles from './SearchForm.module.css'

class SearchForm extends React.Component {
  state = {
    searchData: {
      q: '',
      media_type: 'image,video,audio'
    },
    image: true,
    video: true,
    audio: true
  }

  componentDidMount = () => {
    this.onRefreshPage()
  }

  onRefreshPage = () => {
    const {location, searchMedia, mediasParams} = this.props
    const params = urlParamsToObject(location.search)

    params.q && this.setState({
      searchData: {
        ...this.state.searchData,
        q: params.q
      }
    }, () => {
      this.setMediaTypeParamsToState(params)
      if (params.q !== mediasParams.q) searchMedia(params)
    })
  }

  setMediaTypeParamsToState = (params) => {
    if (params.media_type) {
      this.setState({
        searchData: {
          ...this.state.searchData,
          media_type: params.media_type
        },
        image: false,
        video: false,
        audio: false
      })
      params.media_type.split(',').forEach(type => {
        this.setState({[type]: true})
      })
    }
  }

  onQueryChange = (e) => {
    this.setState({searchData: {...this.state.searchData, q: e.target.value}})
  }

  onImageCheck = (e) => {
    this.setState({image: e.target.checked}, this.reduceMediaType)
  }

  onVideoCheck = (e) => {
    this.setState({video: e.target.checked}, this.reduceMediaType)
  }

  onAudioCheck = (e) => {
    this.setState({audio: e.target.checked}, this.reduceMediaType)
  }

  reduceMediaType = () => {
    const {image, video, audio} = this.state
    const mediaType = `${image ? 'image,' : ''}${video ? 'video,' : ''}${audio ? 'audio,' : ''}`.slice(0, -1)
    this.setState({searchData: {...this.state.searchData, media_type: mediaType}})
  }

  handleKeyPress = (e) => {
    if (e.key === 'Enter') {
      e.preventDefault()
      this.handleSearchMedia()
    }
  }

  handleSearchMedia = () => {
    const {searchMedia, history} = this.props
    const {searchData} = this.state

    searchMedia(searchData)
    history.push(
      `${history.location.pathname}${objectToUrlParams(searchData)}`
    )
  }

  render() {
    const {searchData, video, image, audio} = this.state
    return (
      <div className={styles.wrapper}>
        <div>
          <div className={styles.title}>Search from Nasa</div>
          <Input value={searchData.q}
            onChange={this.onQueryChange}
            onKeyPress={this.handleKeyPress}
            disabled={!video && !image && !audio}
            placeholder="Type something to search..."
          />
        </div>
        <MediaTypeCheckbox
          video={video} image={image} audio={audio}
          onImageCheck={this.onImageCheck}
          onVideoCheck={this.onVideoCheck}
          onAudioCheck={this.onAudioCheck}
        />
      </div>
    )
  }
}

export default withRouter(connect(
  ({medias}) => ({mediasParams: medias.params}),
  (dispatch) => ({
    searchMedia: (data) => dispatch(searchNasaMedia(data))
  })
)(SearchForm))
