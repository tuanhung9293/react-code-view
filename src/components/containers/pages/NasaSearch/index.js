import React from 'react'
import {connect} from 'react-redux'
import {Link} from 'react-router-dom'
import {PAGINATION_UNIT} from 'constants/defaultValues'
import DocumentTitle from 'components/html/DocumentTitle'
import SearchForm from './SearchForm'
import AddCollectionButton from './AddCollectionButton'
import MediasGalaxy from '../../shared/MediasGalaxy'
import PrimaryButton from 'components/ui/buttons/PrimaryButton'
import Loading from 'components/ui/Loading'
import {searchMoreNasaMedia} from 'state/actions/mediaAction'
import styles from './NasaSearch.module.css'

class NasaSearch extends React.Component {
  state = {
    page: 1
  }

  loadMore = () => {
    this.checkMoreSource()
    this.setState({page: this.state.page + 1})
  }

  checkMoreSource = () => {
    const {params, fetchMoreMedias, totalHits, searchResults} = this.props

    if ((totalHits > searchResults.length) && ((this.paginatedSearchResults().length + PAGINATION_UNIT) > searchResults.length)) {
      fetchMoreMedias({...params, page: params.page + 1})
    }
  }

  paginatedSearchResults = () => {
    return this.props.searchResults.slice(0, PAGINATION_UNIT * this.state.page)
  }

  render() {
    const {totalHits, loading, loadingMore} = this.props
    return (
      <React.Fragment>
        <DocumentTitle title="Search" />
        <div className={styles.link}>
          <Link to="/collection">Back to Collection</Link>
        </div>
        <SearchForm />
        {loading && <div className={styles.loading}><Loading /></div>}
        {!loading &&
          <div>
            <MediasGalaxy
              galaxyData={this.paginatedSearchResults()}
              source="medias"
              actionComponent={AddCollectionButton}
              total={totalHits}
            />
            {(this.paginatedSearchResults().length < totalHits) &&
              <div className={styles.loadMore}>
                {loadingMore && <Loading />}
                {!loadingMore && <PrimaryButton onClick={this.loadMore}>Load more</PrimaryButton>}
              </div>
            }
          </div>
        }
      </React.Fragment>
    )
  }
}

export default connect(
  ({medias}) => ({
    searchResults: medias.data,
    totalHits: medias.totalHits,
    params: medias.params,
    loading: medias.loading,
    loadingMore: medias.loadingMore
  }),
  (dispatch) => ({
    fetchMoreMedias: (params) => dispatch(searchMoreNasaMedia(params))
  })
)(NasaSearch)
