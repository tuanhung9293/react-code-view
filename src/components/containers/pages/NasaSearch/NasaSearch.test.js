import React from 'react'
import {mount, shallow} from 'enzyme'
import {MemoryRouter} from 'react-router-dom'
import {Provider} from 'react-redux'
import NasaSearch from './'

import configureStore from 'redux-mock-store'
const mockStore = configureStore()

let store = mockStore({
  medias: {
    data: [{}, {}, {}],
    totalHits: 1002,
    params: {
      q: 'dada',
      mediaType: 'video'
    },
    loading: true,
    loadingMore: false
  }
})
store.dispatch = () => {}


describe('renders', () => {
  it('renders NasaSearch conneted', () => {
    expect(shallow(
      <Provider store={store}>
        <NasaSearch />
      </Provider>
    )).toMatchSnapshot()
  })

  it('renders NasaSearch Component', () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={[{pathname: '/', key: 'testKey'} ]}>
          <NasaSearch />
        </MemoryRouter>
      </Provider>
    )
    expect(wrapper.find(NasaSearch)).toMatchSnapshot()
  })
})



