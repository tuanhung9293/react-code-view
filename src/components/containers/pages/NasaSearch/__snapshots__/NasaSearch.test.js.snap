// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders renders NasaSearch Component 1`] = `
<Connect(NasaSearch)>
  <NasaSearch
    fetchMoreMedias={[Function]}
    loading={true}
    loadingMore={false}
    params={
      Object {
        "mediaType": "video",
        "q": "dada",
      }
    }
    searchResults={
      Array [
        Object {},
        Object {},
        Object {},
      ]
    }
    totalHits={1002}
  >
    <DocumentTitle
      title="Search"
    >
      <SideEffect(DocumentTitle)
        title="Search"
      >
        <DocumentTitle
          title="Search"
        />
      </SideEffect(DocumentTitle)>
    </DocumentTitle>
    <div
      className="link"
    >
      <Link
        replace={false}
        to="/collection"
      >
        <a
          href="/collection"
          onClick={[Function]}
        >
          Back to Collection
        </a>
      </Link>
    </div>
    <withRouter(Connect(SearchForm))>
      <Route>
        <Connect(SearchForm)
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "testKey",
                  "pathname": "/",
                  "search": "",
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "testKey",
                "pathname": "/",
                "search": "",
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/",
              "search": "",
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <SearchForm
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "key": "testKey",
                    "pathname": "/",
                    "search": "",
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 0,
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "key": "testKey",
                  "pathname": "/",
                  "search": "",
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "key": "testKey",
                "pathname": "/",
                "search": "",
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            mediasParams={
              Object {
                "mediaType": "video",
                "q": "dada",
              }
            }
            searchMedia={[Function]}
          >
            <div
              className="wrapper"
            >
              <div>
                <div
                  className="title"
                >
                  Search from Nasa
                </div>
                <Input
                  disabled={false}
                  onChange={[Function]}
                  onKeyPress={[Function]}
                  placeholder="Type something to search..."
                  value=""
                >
                  <input
                    className="input"
                    disabled={false}
                    onChange={[Function]}
                    onKeyPress={[Function]}
                    placeholder="Type something to search..."
                    type="text"
                    value=""
                  />
                </Input>
              </div>
              <MediaTypeCheckbox
                audio={true}
                image={true}
                onAudioCheck={[Function]}
                onImageCheck={[Function]}
                onVideoCheck={[Function]}
                video={true}
              >
                <div
                  className="wrapper"
                >
                  <div
                    className="inlineFlex"
                  >
                    <label
                      className="checkbox"
                    >
                      <Checkbox
                        checked={true}
                        onChange={[Function]}
                      >
                        <span
                          className="wrapper"
                        >
                          <input
                            checked={true}
                            className="input"
                            onChange={[Function]}
                            type="checkbox"
                          />
                          <span
                            className="display"
                          />
                        </span>
                      </Checkbox>
                      Image
                    </label>
                    <label
                      className="checkbox"
                    >
                      <Checkbox
                        checked={true}
                        onChange={[Function]}
                      >
                        <span
                          className="wrapper"
                        >
                          <input
                            checked={true}
                            className="input"
                            onChange={[Function]}
                            type="checkbox"
                          />
                          <span
                            className="display"
                          />
                        </span>
                      </Checkbox>
                      Video
                    </label>
                    <label
                      className="checkbox"
                    >
                      <Checkbox
                        checked={true}
                        onChange={[Function]}
                      >
                        <span
                          className="wrapper"
                        >
                          <input
                            checked={true}
                            className="input"
                            onChange={[Function]}
                            type="checkbox"
                          />
                          <span
                            className="display"
                          />
                        </span>
                      </Checkbox>
                      Audio
                    </label>
                  </div>
                </div>
              </MediaTypeCheckbox>
            </div>
          </SearchForm>
        </Connect(SearchForm)>
      </Route>
    </withRouter(Connect(SearchForm))>
    <div
      className="loading"
    >
      <Loading>
        <div
          className="spinner"
        />
      </Loading>
    </div>
  </NasaSearch>
</Connect(NasaSearch)>
`;

exports[`renders renders NasaSearch conneted 1`] = `<Connect(NasaSearch) />`;
